### PR TITLE
crypto review action: in_out is a BytesMut instance should be scrubbed before returning

### LINKED ISSUE #585: crypto review action: in_out is a BytesMut instance should be scrubbed before returning
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/longlon1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/longlon1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
</head>

<body link="#0563C1" vlink="#954F72">

spdmlib/crypto/spdm_ring | encrypt() | aead_impl.rs | 60 | 1 | in_out is a BytesMut instance   that contains sensitive data (e.g. ciphertext). This should be scrubbed   before returning. The same issue exists for decrypt().
-- | -- | -- | -- | -- | --

</body>

</html>

https://github.com/jyao1/rust-spdm/blob/81a2bc40a14be7dcb7b90add073e6f64e3eddf9d/spdmlib/src/crypto/spdm_ring/aead_impl.rs#L54-L71

https://github.com/jyao1/rust-spdm/blob/81a2bc40a14be7dcb7b90add073e6f64e3eddf9d/spdmlib/src/crypto/spdm_ring/aead_impl.rs#L111-L128

### PR DESCRIPTION
**Pull request by longlongyang**
fix #585

### CODE REVIEW

File: spdmlib/src/responder/key_exchange_rsp.rs [hunk 1]
```diff
@@ -516,7 +516,7 @@ mod tests_responder {
                 .ok()
                 .unwrap();
         let public_key_old = private_key.compute_public_key().ok().unwrap();
-        let public_key = BytesMut::from(&public_key_old.as_ref()[1..]);
+        let public_key = BytesMutStrubbed::from(&public_key_old.as_ref()[1..]);
```
Comment (jyao1) [inline spdmlib/src/responder/key_exchange_rsp.rs:519]: you dont need this for public_key

File: spdmlib/src/crypto/bytes_mut_scrubbed.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
+use bytes::{buf::IntoIter, Buf, BufMut, Bytes, BytesMut};
```
Comment (jyao1) [inline spdmlib/src/crypto/bytes_mut_scrubbed.rs:1]: please add license header.

File: spdmlib/src/crypto/spdm_ring/aead_impl.rs [hunk 3]
```diff
@@ -2,9 +2,9 @@
 //
 // SPDX-License-Identifier: BSD-2-Clause-Patent
 
+use crate::crypto::bytes_mut_scrubbed::BytesMutStrubbed;
```
Comment (jyao1) [inline spdmlib/src/crypto/spdm_ring/aead_impl.rs:5]: this is for ring only.
Do we need this for mbedtls binding?

File: spdmlib/src/responder/key_exchange_rsp.rs (hunk 1 above)
Comment (longlongyang) [reply spdmlib/src/responder/key_exchange_rsp.rs]: > you dont need this for public_key

Yes, updated. So now the struct use BytesMut are SpdmDigestStruct, SpdmSignatureStruct and SpdmDheExchangeStruct.
The struct use BytesMutStrubbed are SpdmHKDFKeyStruct, SpdmDheFinalKeyStruct, SpdmAeadKeyStruct and SpdmAeadIvStruct.

File: spdmlib/src/crypto/bytes_mut_scrubbed.rs (hunk 2 above)
Comment (longlongyang) [reply spdmlib/src/crypto/bytes_mut_scrubbed.rs:5]: > please add license header.

updated.

File: spdmlib/src/crypto/spdm_ring/aead_impl.rs (hunk 3 above)
Comment (longlongyang) [reply spdmlib/src/crypto/spdm_ring/aead_impl.rs:5]: > this is for ring only. Do we need this for mbedtls binding?

The implementation of mbedtls binding doesn't use BytesMut at all. Use the raw u8 slice.
How about let's create a new issue to track this?

Comment (jyao1) [reply spdmlib/src/crypto/spdm_ring/aead_impl.rs:5]: yes, let's create another issue

Comment (longlongyang) [reply spdmlib/src/crypto/spdm_ring/aead_impl.rs:5]: > yes, let's create another issue

created here https://github.com/jyao1/rust-spdm/issues/605

**Pull request merged**